aten/src/ATen/native/native_functions.yaml

```diff
@@ -2462,6 +2462,7 @@
 
 - func: numel(Tensor self) -> int
   variants: function, method
+  requires_tensor: True
```

Comment:
This is needed for `numel_override` in MSNPU extension backend ops to work. Here is the reasoning:

Before this patch, when we run the following code:
```
torch.zeros(5, 5, device='msnpu').numel()
```
, the following code path is executed:
```
#1  0x00007fffe2f2abc0 in at::MSNPUType::numel (this=0x555555f03870, self=...)
#2  0x00007fffe86eab91 in at::Tensor::numel (this=0x55555592e128) at ../aten/src/ATen/core/TensorMethods.h:789
#3  0x00007fffe05f87b1 in torch::autograd::Variable::Impl::numel (this=0x55555592e040) at ../torch/csrc/autograd/variable.cpp:47
#4  0x00007fffe2c29f4b in at::native::numel (self=...) at ../aten/src/ATen/native/TensorShape.cpp:805
#5  0x00007fffe2f2abc0 in at::TypeDefault::numel (this=0x555555f03f30, self=...) at aten/src/ATen/TypeDefault.cpp:4324
#6  0x00007fffe00bdc2d in torch::autograd::VariableType::numel (this=0x555555f03f30, self=...) at ../torch/csrc/autograd/generated/VariableType_0.cpp:13377
#7  0x00007fffe86eab91 in at::Tensor::numel (this=0x7ffff6d8a058) at ../aten/src/ATen/core/TensorMethods.h:789
#8 0x00007fffe87c679a in torch::autograd::dispatch_numel (self=...) at ../torch/csrc/autograd/generated/python_variable_methods_dispatch.h:1410
#9 0x00007fffe87866cf in torch::autograd::THPVariable_numel (self_=0x7ffff6d8a048, args=0x0) at ../torch/csrc/autograd/generated/python_variable_methods.cpp:3682
```
The implementation of `at::native::numel` in `aten/src/ATen/native/TensorShape.cpp:805` is `self.unsafeGetTensorImpl()->numel()`.  In this case, `at::native::numel` calls into `torch::autograd::Variable::Impl::numel`, which calls `data_.numel()` and thus dispatches to `at::MSNPUType::numel` correctly.

After this patch, `Variable::Impl` is merged into `TensorImpl`, and as a result, the following code path is executed:
```
#1  0x00007fffe1922af7 in c10::TensorImpl::numel (this=0x5555562f9670) at ../c10/core/TensorImpl.h:375
#2  0x00007fffe2c29f4b in at::native::numel (self=...) at ../aten/src/ATen/native/TensorShape.cpp:805
#3  0x00007fffe2f2abc0 in at::TypeDefault::numel (this=0x555555f03f30, self=...) at aten/src/ATen/TypeDefault.cpp:4324
#4  0x00007fffe00bdc2d in torch::autograd::VariableType::numel (this=0x555555f03f30, self=...) at ../torch/csrc/autograd/generated/VariableType_0.cpp:13377
#5  0x00007fffe86eab91 in at::Tensor::numel (this=0x7ffff6d8a058) at ../aten/src/ATen/core/TensorMethods.h:789
#6 0x00007fffe87c679a in torch::autograd::dispatch_numel (self=...) at ../torch/csrc/autograd/generated/python_variable_methods_dispatch.h:1410
#7 0x00007fffe87866cf in torch::autograd::THPVariable_numel (self_=0x7ffff6d8a048, args=0x0) at ../torch/csrc/autograd/generated/python_variable_methods.cpp:3682
```

Since the implementation of `at::native::numel` is still `self.unsafeGetTensorImpl()->numel()`, in this case `at::native::numel` directly calls into `c10::TensorImpl::numel`, which ignores the `at::MSNPUType::numel` backend extension operator, and is incorrect behavior.

The fix suggested here is to let `torch::autograd::VariableType::numel` use `baseType` dispatch instead of TypeDefault dispatch, so that the code path will look like:
```
#1  0x00007fffe2f2abc0 in at::MSNPUType::numel (this=0x555555f03870, self=...)
#2  0x00007fffe00bdc2d in torch::autograd::VariableType::numel (this=0x555555f03f30, self=...) at ../torch/csrc/autograd/generated/VariableType_0.cpp:13377
#3  0x00007fffe86eab91 in at::Tensor::numel (this=0x7ffff6d8a058) at ../aten/src/ATen/core/TensorMethods.h:789
#4 0x00007fffe87c679a in torch::autograd::dispatch_numel (self=...) at ../torch/csrc/autograd/generated/python_variable_methods_dispatch.h:1410
#5 0x00007fffe87866cf in torch::autograd::THPVariable_numel (self_=0x7ffff6d8a048, args=0x0) at ../torch/csrc/autograd/generated/python_variable_methods.cpp:3682
```

Reply:
The current MSNPU test actually overrides the `numel()` function in: https://github.com/pytorch/pytorch/blob/236c2b238724364129e9670b42a555e976fd635b/test/cpp_extensions/msnpu_extension.cpp#L53-L56 and it expects to be able to register this new `numel()` as an extension backend op: https://github.com/pytorch/pytorch/blob/236c2b238724364129e9670b42a555e976fd635b/test/cpp_extensions/msnpu_extension.cpp#L84-L86
If we let MSNPUType use the base `numel()` instead of allowing it to override the `numel()` function, the call site of `msnpu_tensor.numel()` will receive an incorrect returned value, and this MSNPUType test will fail:
https://github.com/pytorch/pytorch/blob/236c2b238724364129e9670b42a555e976fd635b/test/test_cpp_extensions.py#L673-L674